File: packages/interop/python/wasm_loader.py
"""
WASM Plugin Loader for elizaOS

This module provides utilities for loading WASM plugins into the Python runtime
via wasmtime or similar WASM runtimes.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Callable, Awaitable

# Try to import wasmtime, provide fallback if not available
try:
    import wasmtime
    HAS_WASMTIME = True
except ImportError:
    HAS_WASMTIME = False
    wasmtime = None  # type: ignore

from elizaos.types.plugin import Plugin
from elizaos.types.memory import Memory
from elizaos.types.state import State
from elizaos.types.components import (
    Action,
    ActionResult,
    Provider,
    ProviderResult,
    Evaluator,
    HandlerOptions,
)


class WasmPluginLoader:
    """
    WASM plugin loader for elizaOS.

    Loads Rust or TypeScript plugins compiled to WASM and adapts them
    to the Python Plugin interface.

    Requires wasmtime-py: pip install wasmtime

    The WASM plugin must export these functions:
    - get_manifest() -> ptr (JSON string)
    - init(config_ptr: i32, config_len: i32)
    - validate_action(name_ptr, name_len, memory_ptr, memory_len, state_ptr, state_len) -> i32
    - invoke_action(name_ptr, name_len, memory_ptr, memory_len, state_ptr, state_len, options_ptr, options_len) -> ptr
    - get_provider(name_ptr, name_len, memory_ptr, memory_len, state_ptr, state_len) -> ptr
    - validate_evaluator(name_ptr, name_len, memory_ptr, memory_len, state_ptr, state_len) -> i32
    - invoke_evaluator(name_ptr, name_len, memory_ptr, memory_len, state_ptr, state_len) -> ptr
    - alloc(size: i32) -> ptr
    - dealloc(ptr: i32, size: i32)
    """

    def __init__(
        self,
        wasm_path: str | Path,
        *,
        max_module_bytes: int | None = None,
        max_memory_bytes: int | None = None,
        fuel: int | None = None,
        max_string_bytes: int = 1_000_000,
    ) -> None:
        """
        Initialize the WASM plugin loader.

        Args:
            wasm_path: Path to the WASM file.

        Raises:
            ImportError: If wasmtime is not installed.
            FileNotFoundError: If the WASM file doesn't exist.
        """
        if not HAS_WASMTIME:
            raise ImportError(
                "wasmtime is required for WASM plugin loading. "
                "Install with: pip install wasmtime"
            )

        self.wasm_path = Path(wasm_path)
        if not self.wasm_path.exists():
            raise FileNotFoundError(f"WASM file not found: {wasm_path}")

        if max_module_bytes is not None:
            size = self.wasm_path.stat().st_size
            if size > max_module_bytes:
                raise ValueError(
                    f"WASM module too large ({size} bytes > {max_module_bytes} bytes)"
                )

        self.max_memory_bytes = max_memory_bytes
        self.max_string_bytes = max_string_bytes

        # Initialize wasmtime engine and store
        if fuel is not None:
            config = wasmtime.Config()
            config.consume_fuel = True
            self.engine = wasmtime.Engine(config)
        else:
            self.engine = wasmtime.Engine()
        self.store = wasmtime.Store(self.engine)
        if fuel is not None:
            self.store.add_fuel(fuel)
        self.linker = wasmtime.Linker(self.engine)

        # Load and instantiate the module
        self._setup_imports()
        self._load_module()
        self.manifest: dict[str, Any] | None = None

    def _setup_imports(self) -> None:
        """Set up the import functions for the WASM module."""
        # WASI imports (minimal stubs)
        wasi_config = wasmtime.WasiConfig()
        wasi_config.inherit_stdout()
        wasi_config.inherit_stderr()
        self.store.set_wasi(wasi_config)
        self.linker.define_wasi()

        # Environment imports for console logging
        @wasmtime.Func(self.store, wasmtime.FuncType([wasmtime.ValType.i32(), wasmtime.ValType.i32()], []))
        def console_log(ptr: int, len_: int) -> None:
            data = self._read_memory(ptr, len_)
            print(data.decode("utf-8"))

        @wasmtime.Func(self.store, wasmtime.FuncType([wasmtime.ValType.i32(), wasmtime.ValType.i32()], []))
        def console_error(ptr: int, len_: int) -> None:
            data = self._read_memory(ptr, len_)
            print(f"[ERROR] {data.decode('utf-8')}")

        self.linker.define(self.store, "env", "console_log", console_log)
        self.linker.define(self.store, "env", "console_error", console_error)

    def _load_module(self) -> None:
        """Load and instantiate the WASM module."""
        with open(self.wasm_path, "rb") as f:
            wasm_bytes = f.read()

        self.module = wasmtime.Module(self.engine, wasm_bytes)
        self.instance = self.linker.instantiate(self.store, self.module)

        # Get memory export
        memory_export = self.instance.exports(self.store).get("memory")
        if memory_export is None:
            raise RuntimeError("WASM module does not export 'memory'")
        self.memory = memory_export

        if self.max_memory_bytes is not None:
            mem_data = self.memory.data_ptr(self.store)
            if len(mem_data) > self.max_memory_bytes:
                raise RuntimeError(
                    f"WASM memory too large ({len(mem_data)} bytes > {self.max_memory_bytes} bytes)"
                )

    def _read_memory(self, ptr: int, length: int) -> bytes:
        """Read bytes from WASM memory."""
        data = self.memory.data_ptr(self.store)
        return bytes(data[ptr:ptr + length])

    def _write_memory(self, ptr: int, data: bytes) -> None:
        """Write bytes to WASM memory."""
        mem_data = self.memory.data_ptr(self.store)
        for i, byte in enumerate(data):
            mem_data[ptr + i] = byte

    def _read_string(self, ptr: int) -> str:
        """Read a null-terminated string from WASM memory."""
        mem_data = self.memory.data_ptr(self.store)
        end = ptr
        limit = ptr + self.max_string_bytes
        while end < len(mem_data) and end < limit and mem_data[end] != 0:
            end += 1
        if end >= len(mem_data) or end >= limit:
            raise RuntimeError("WASM string exceeded maximum length or memory bounds")
        return bytes(mem_data[ptr:end]).decode("utf-8")

    def _call_with_string(self, func_name: str, *strings: str) -> str | None:
        """Call a WASM function with string arguments and get string result."""
        exports = self.instance.exports(self.store)
        func = exports.get(func_name)
        if func is None:
            raise RuntimeError(f"WASM function '{func_name}' not found")

        alloc = exports.get("alloc")
        dealloc = exports.get("dealloc")
        if alloc is None or dealloc is None:
            raise RuntimeError("WASM module must export 'alloc' and 'dealloc'")

        # Allocate and write each string
        ptrs_and_lens: list[tuple[int, int]] = []
        for s in strings:
            encoded = s.encode("utf-8")
            ptr = alloc(self.store, len(encoded))
            self._write_memory(ptr, encoded)
            ptrs_and_lens.append((ptr, len(encoded)))

        # Build arguments (alternating ptr, len for each string)
        args = []
        for ptr, length in ptrs_and_lens:
            args.extend([ptr, length])

        # Call the function
        result_ptr = func(self.store, *args)

        # Cleanup input strings
        for ptr, length in ptrs_and_lens:
            dealloc(self.store, ptr, length)

        if result_ptr == 0:
            return None

        return self._read_string(result_ptr)

    def get_manifest(self) -> dict[str, Any]:
        """Get the plugin manifest from the WASM module."""
        if self.manifest is not None:
            return self.manifest

        exports = self.instance.exports(self.store)
        get_manifest_fn = exports.get("get_manifest")
        if get_manifest_fn is None:
            raise RuntimeError("WASM module must export 'get_manifest'")

        result_ptr = get_manifest_fn(self.store)
        if result_ptr == 0:
            raise RuntimeError("get_manifest returned null")

        json_str = self._read_string(result_ptr)
        self.manifest = json.loads(json_str)
        return self.manifest

    def init(self, config: dict[str, str]) -> None:
        """Initialize the plugin with configuration."""
        exports = self.instance.exports(self.store)
        init_fn = exports.get("init")
        if init_fn is None:
            return  # init is optional

        alloc = exports.get("alloc")
        if alloc is None:
            return

        config_json = json.dumps(config).encode("utf-8")
        ptr = alloc(self.store, len(config_json))
        self._write_memory(ptr, config_json)
        init_fn(self.store, ptr, len(config_json))

    def validate_action(self, name: str, memory: Memory, state: State | None) -> bool:
        """Validate an action."""
        result = self._call_validate("validate_action", name, memory, state)
        return result != 0

    def invoke_action(
        self, name: str, memory: Memory, state: State | None, options: dict[str, Any] | None
    ) -> ActionResult:
        """Invoke an action."""
        memory_json = json.dumps(memory.model_dump() if hasattr(memory, "model_dump") else memory)
        state_json = json.dumps(
            state.model_dump() if state and hasattr(state, "model_dump") else state
        )
        options_json = json.dumps(options or {})

        result_json = self._call_with_string(
            "invoke_action", name, memory_json, state_json, options_json
        )
        if not result_json:
            return ActionResult(success=False, error="No result from WASM plugin")

        result_data = json.loads(result_json)
        return ActionResult(**result_data)

    def get_provider(self, name: str, memory: Memory, state: State) -> ProviderResult:
        """Get provider data."""
        memory_json = json.dumps(memory.model_dump() if hasattr(memory, "model_dump") else memory)
        state_json = json.dumps(state.model_dump() if hasattr(state, "model_dump") else state)

        result_json = self._call_with_string("get_provider", name, memory_json, state_json)
        if not result_json:
            return ProviderResult()

        result_data = json.loads(result_json)
        return ProviderResult(**result_data)

    def validate_evaluator(self, name: str, memory: Memory, state: State | None) -> bool:
        """Validate an evaluator."""
        result = self._call_validate("validate_evaluator", name, memory, state)
        return result != 0

    def invoke_evaluator(self, name: str, memory: Memory, state: State | None) -> ActionResult | None:
        """Invoke an evaluator."""
        memory_json = json.dumps(memory.model_dump() if hasattr(memory, "model_dump") else memory)
        state_json = json.dumps(
            state.model_dump() if state and hasattr(state, "model_dump") else state
        )

        result_json = self._call_with_string("invoke_evaluator", name, memory_json, state_json)
        if not result_json or result_json == "null":
            return None

        result_data = json.loads(result_json)
        return ActionResult(**result_data)

    def _call_validate(
        self, func_name: str, name: str, memory: Memory, state: State | None
    ) -> int:
        """Call a validation function."""
        exports = self.instance.exports(self.store)
        func = exports.get(func_name)
        alloc = exports.get("alloc")
        dealloc = exports.get("dealloc")

        if func is None or alloc is None or dealloc is None:
            return 0

        name_bytes = name.encode("utf-8")
        memory_json = json.dumps(
            memory.model_dump() if hasattr(memory, "model_dump") else memory
        ).encode("utf-8")
        state_json = json.dumps(
            state.model_dump() if state and hasattr(state, "model_dump") else state
        ).encode("utf-8")

        name_ptr = alloc(self.store, len(name_bytes))
        mem_ptr = alloc(self.store, len(memory_json))
        state_ptr = alloc(self.store, len(state_json))

        self._write_memory(name_ptr, name_bytes)
        self._write_memory(mem_ptr, memory_json)
        self._write_memory(state_ptr, state_json)

        result = func(
            self.store,
            name_ptr, len(name_bytes),
            mem_ptr, len(memory_json),
            state_ptr, len(state_json),
        )

        dealloc(self.store, name_ptr, len(name_bytes))
        dealloc(self.store, mem_ptr, len(memory_json))
        dealloc(self.store, state_ptr, len(state_json))

        return result


def load_wasm_plugin(
    wasm_path: str | Path,
    manifest_path: str | Path | None = None,
    *,
    max_module_bytes: int | None = None,
    max_memory_bytes: int | None = None,
    fuel: int | None = None,
) -> Plugin:
    """
    Load a WASM plugin and return an elizaOS Plugin interface.

    Args:
        wasm_path: Path to the WASM file.
        manifest_path: Optional path to external manifest JSON (uses embedded if not provided).

    Returns:
        elizaOS Plugin instance.
    """
    loader = WasmPluginLoader(
        wasm_path,
        max_module_bytes=max_module_bytes,
        max_memory_bytes=max_memory_bytes,
        fuel=fuel,
    )

    # Get manifest
    if manifest_path:
        with open(manifest_path) as f:
            manifest = json.load(f)
    else:
        manifest = loader.get_manifest()

    # Create action wrappers
    actions: list[Action] = []
    for action_def in manifest.get("actions", []):
        action_name = action_def["name"]

        def make_validate(name: str) -> Callable[..., Awaitable[bool]]:
            async def validate(runtime: Any, message: Memory, state: State | None) -> bool:
                return loader.validate_action(name, message, state)
            return validate

        def make_handler(name: str) -> Callable[..., Awaitable[ActionResult | None]]:
            async def handler(
                runtime: Any,
                message: Memory,
                state: State | None,
                options: HandlerOptions | None,
                callback: Any,
                responses: Any,
            ) -> ActionResult | None:
                return loader.invoke_action(
                    name, message, state, options.model_dump() if options else None
                )
            return handler

        validate_fn = make_validate(action_name)
        handler_fn = make_handler(action_name)

        actions.append(
            Action(
                name=action_name,
                description=action_def.get("description", ""),
                similes=action_def.get("similes"),
                examples=action_def.get("examples"),
                validate=validate_fn,  # type: ignore
                handler=handler_fn,  # type: ignore
            )
        )

    # Create provider wrappers
    providers: list[Provider] = []
    for provider_def in manifest.get("providers", []):
        provider_name = provider_def["name"]

        def make_get(name: str) -> Callable[..., Awaitable[ProviderResult]]:
            async def get(runtime: Any, message: Memory, state: State) -> ProviderResult:
                return loader.get_provider(name, message, state)
            return get

        get_fn = make_get(provider_name)

        providers.append(
            Provider(
                name=provider_name,
                description=provider_def.get("description"),
                dynamic=provider_def.get("dynamic"),
                position=provider_def.get("position"),
                private=provider_def.get("private"),
                get=get_fn,  # type: ignore
            )
        )

    # Create evaluator wrappers
    evaluators: list[Evaluator] = []
    for eval_def in manifest.get("evaluators", []):
        eval_name = eval_def["name"]

        def make_eval_validate(name: str) -> Callable[..., Awaitable[bool]]:
            async def validate(runtime: Any, message: Memory, state: State | None) -> bool:
                return loader.validate_evaluator(name, message, state)
            return validate

        def make_eval_handler(name: str) -> Callable[..., Awaitable[ActionResult | None]]:
            async def handler(
                runtime: Any,
                message: Memory,
                state: State | None,
                options: HandlerOptions | None,
                callback: Any,
                responses: Any,
            ) -> ActionResult | None:
                return loader.invoke_evaluator(name, message, state)
            return handler

        validate_fn = make_eval_validate(eval_name)
        handler_fn = make_eval_handler(eval_name)

        evaluators.append(
            Evaluator(
                name=eval_name,
                description=eval_def.get("description", ""),
                always_run=eval_def.get("alwaysRun"),
                similes=eval_def.get("similes"),
                examples=[],
                validate=validate_fn,  # type: ignore
                handler=handler_fn,  # type: ignore
            )
        )

    # Create init function
    async def init(config: dict[str, str], runtime: Any) -> None:
        loader.init(config)

    return Plugin(
        name=manifest["name"],
        description=manifest["description"],
        init=init,
        config=manifest.get("config"),
        dependencies=manifest.get("dependencies"),
        actions=actions if actions else None,
        providers=providers if providers else None,
        evaluators=evaluators if evaluators else None,
    )


def validate_wasm_plugin(wasm_path: str | Path) -> dict[str, Any]:
    """
    Validate a WASM plugin without fully loading it.

    Args:
        wasm_path: Path to the WASM file.

    Returns:
        Dict with 'valid', 'manifest', and optionally 'error' keys.
    """
    try:
        loader = WasmPluginLoader(wasm_path)
        manifest = loader.get_manifest()

        if not manifest.get("name") or not manifest.get("description"):
            return {"valid": False, "error": "Missing required manifest fields"}

        return {"valid": True, "manifest": manifest}
    except Exception as e:
        return {"valid": False, "error": str(e)}

File: packages/interop/typescript/wasm-loader.ts
/**
 * WASM Plugin Loader for elizaOS
 *
 * Loads Rust (or other) plugins compiled to WebAssembly and adapts them
 * to the TypeScript Plugin interface.
 */

import type {
  Action,
  ActionResult,
  EvaluationExample,
  Evaluator,
  HandlerCallback,
  HandlerOptions,
  IAgentRuntime,
  Memory,
  Plugin,
  Provider,
  ProviderResult,
  ProviderValue,
  State,
} from "@elizaos/core";
import { logger } from "@elizaos/core";

import type {
  ActionResultPayload,
  PluginManifest,
  ProviderResultPayload,
  WasmPluginExports,
  WasmPluginInstance,
} from "./types";

/**
 * Options for loading a WASM plugin
 */
export interface WasmLoaderOptions {
  /** Path or URL to the WASM file */
  wasmPath: string;
  /** Optional path to plugin.json manifest (auto-detected if not provided) */
  manifestPath?: string;
  /** Import object for WASM instantiation */
  imports?: WebAssembly.Imports;
  /** Maximum allowed WASM binary size in bytes. */
  maxWasmBytes?: number;
  /** Maximum allowed initial memory size in bytes (post-instantiation). */
  maxMemoryBytes?: number;
}

/**
 * Text encoder/decoder for string passing
 */
const encoder = new TextEncoder();
const decoder = new TextDecoder();

/**
 * Load a WASM plugin and return an elizaOS Plugin interface
 */
export async function loadWasmPlugin(
  options: WasmLoaderOptions,
): Promise<Plugin> {
  const { wasmPath, manifestPath } = options;

  // Load the WASM module
  const wasmInstance = await loadWasmModule(wasmPath, options);

  // Get the manifest from the WASM module or external file
  let manifest: PluginManifest;
  if (manifestPath) {
    manifest = await loadManifest(manifestPath);
  } else {
    // Try to get manifest from WASM exports
    const manifestJson = wasmInstance.exports.get_manifest();
    manifest = JSON.parse(manifestJson);
  }

  // Create the plugin adapter
  return createPluginFromWasm(manifest, wasmInstance);
}

/**
 * Load a WASM module and instantiate it
 */
async function loadWasmModule(
  wasmPath: string,
  options: Pick<
    WasmLoaderOptions,
    "imports" | "maxWasmBytes" | "maxMemoryBytes"
  >,
): Promise<WasmPluginInstance> {
  // Default imports for WASM
  const defaultImports: WebAssembly.Imports = {
    env: {
      // Console logging
      console_log: (_ptr: number, _len: number) => {
        // Will be resolved once we have memory
      },
      console_error: (_ptr: number, _len: number) => {
        // Will be resolved once we have memory
      },
      // Abort handler
      abort: (_msg: number, file: number, line: number, column: number) => {
        throw new Error(`WASM abort at ${file}:${line}:${column}`);
      },
    },
    wasi_snapshot_preview1: {
      // Minimal WASI stubs for compatibility
      proc_exit: (code: number) => {
        throw new Error(`WASM exit with code ${code}`);
      },
      fd_write: () => 0,
      fd_read: () => 0,
      fd_close: () => 0,
      fd_seek: () => 0,
      environ_get: () => 0,
      environ_sizes_get: () => 0,
      clock_time_get: () => 0,
      random_get: (_buf: number, _len: number) => {
        // Fill with random bytes
        return 0;
      },
    },
  };

  const imports = { ...defaultImports, ...options.imports };

  // Determine if we're in Node.js or browser
  const isNode =
    typeof globalThis.process !== "undefined" &&
    globalThis.process.versions &&
    globalThis.process.versions.node;

  let wasmModule: WebAssembly.Module;
  if (isNode) {
    // Node.js: read file
    const fs = await import("node:fs/promises");
    if (typeof options.maxWasmBytes === "number") {
      const st = await fs.stat(wasmPath);
      if (st.size > options.maxWasmBytes) {
        throw new Error(
          `WASM binary too large (${st.size} bytes > ${options.maxWasmBytes} bytes)`,
        );
      }
    }
    const wasmBuffer = await fs.readFile(wasmPath);
    if (
      typeof options.maxWasmBytes === "number" &&
      wasmBuffer.byteLength > options.maxWasmBytes
    ) {
      throw new Error(
        `WASM binary too large (${wasmBuffer.byteLength} bytes > ${options.maxWasmBytes} bytes)`,
      );
    }
    wasmModule = await WebAssembly.compile(wasmBuffer);
  } else {
    // Browser: fetch
    const response = await fetch(wasmPath);
    const wasmBuffer = await response.arrayBuffer();
    if (
      typeof options.maxWasmBytes === "number" &&
      wasmBuffer.byteLength > options.maxWasmBytes
    ) {
      throw new Error(
        `WASM binary too large (${wasmBuffer.byteLength} bytes > ${options.maxWasmBytes} bytes)`,
      );
    }
    wasmModule = await WebAssembly.compile(wasmBuffer);
  }

  const instance = await WebAssembly.instantiate(wasmModule, imports);

  // Set up console logging now that we have memory
  const memory = instance.exports.memory as WebAssembly.Memory;
  if (
    typeof options.maxMemoryBytes === "number" &&
    memory.buffer.byteLength > options.maxMemoryBytes
  ) {
    throw new Error(
      `WASM memory too large (${memory.buffer.byteLength} bytes > ${options.maxMemoryBytes} bytes)`,
    );
  }

  if (imports.env) {
    const env = imports.env as Record<string, unknown>;
    env.console_log = (ptr: number, len: number) => {
      const bytes = new Uint8Array(memory.buffer, ptr, len);
      logger.info(
        { src: "interop:wasm", event: "interop.wasm.stdout", stream: "stdout" },
        decoder.decode(bytes),
      );
    };
    env.console_error = (ptr: number, len: number) => {
      const bytes = new Uint8Array(memory.buffer, ptr, len);
      logger.error(
        { src: "interop:wasm", event: "interop.wasm.stderr", stream: "stderr" },
        decoder.decode(bytes),
      );
    };
  }

  // Provide secure randomness for WASI random_get if present
  if (imports.wasi_snapshot_preview1) {
    const wasi = imports.wasi_snapshot_preview1 as Record<string, unknown>;
    wasi.random_get = (buf: number, len: number) => {
      const view = new Uint8Array(memory.buffer, buf, len);
      const cryptoObj = globalThis.crypto;
      if (!cryptoObj || typeof cryptoObj.getRandomValues !== "function") {
        throw new Error(
          "No secure random source available for WASI random_get",
        );
      }
      cryptoObj.getRandomValues(view);
      return 0;
    };
  }

  // Type guard to validate exports match WasmPluginExports interface
  const exports = instance.exports as WebAssembly.Exports;
  const typedExports: WasmPluginExports = {
    get_manifest: exports.get_manifest as () => string,
    init: exports.init as (config_json: string) => void,
    validate_action: exports.validate_action as (
      action: string,
      memory_json: string,
      state_json: string,
    ) => boolean,
    invoke_action: exports.invoke_action as (
      action: string,
      memory_json: string,
      state_json: string,
      options_json: string,
    ) => string,
    get_provider: exports.get_provider as (
      provider: string,
      memory_json: string,
      state_json: string,
    ) => string,
    validate_evaluator: exports.validate_evaluator as (
      evaluator: string,
      memory_json: string,
      state_json: string,
    ) => boolean,
    invoke_evaluator: exports.invoke_evaluator as (
      evaluator: string,
      memory_json: string,
      state_json: string,
    ) => string,
    handle_route: exports.handle_route as (
      path: string,
      method: string,
      request_json: string,
    ) => string,
    alloc: exports.alloc as (size: number) => number,
    dealloc: exports.dealloc as (ptr: number, size: number) => void,
  };

  return {
    exports: typedExports,
    memory: { buffer: memory.buffer },
  };
}

/**
 * Load manifest from external JSON file
 */
async function loadManifest(manifestPath: string): Promise<PluginManifest> {
  const isNode =
    typeof globalThis.process !== "undefined" &&
    globalThis.process.versions &&
    globalThis.process.versions.node;

  if (isNode) {
    const fs = await import("node:fs/promises");
    const content = await fs.readFile(manifestPath, "utf-8");
    return JSON.parse(content);
  } else {
    const response = await fetch(manifestPath);
    return response.json();
  }
}

/**
 * Helper to pass a string to WASM and get result
 */
function _callWasmWithString(
  instance: WasmPluginInstance,
  fn: (...args: number[]) => number,
  ...strings: string[]
): string {
  const { exports, memory } = instance;
  const ptrs: number[] = [];
  const lens: number[] = [];

  // Allocate and write strings
  for (const str of strings) {
    const bytes = encoder.encode(str);
    const ptr = exports.alloc(bytes.length);
    const view = new Uint8Array(memory.buffer, ptr, bytes.length);
    view.set(bytes);
    ptrs.push(ptr);
    lens.push(bytes.length);
  }

  // Call the function
  const args: number[] = [];
  for (let i = 0; i < strings.length; i++) {
    args.push(ptrs[i], lens[i]);
  }
  const resultPtr = fn(...args);

  // Read result (assuming null-terminated or length-prefixed)
  // For now, assume it returns a pointer to a null-terminated string
  let resultLen = 0;
  const view = new Uint8Array(memory.buffer);
  while (
    view[resultPtr + resultLen] !== 0 &&
    resultPtr + resultLen < view.length
  ) {
    resultLen++;
  }
  const result = decoder.decode(
    new Uint8Array(memory.buffer, resultPtr, resultLen),
  );

  // Deallocate input strings
  for (let i = 0; i < strings.length; i++) {
    exports.dealloc(ptrs[i], lens[i]);
  }

  return result;
}

/**
 * Create a Plugin from a WASM instance
 */
function createPluginFromWasm(
  manifest: PluginManifest,
  instance: WasmPluginInstance,
): Plugin {
  const { exports } = instance;

  // Create action wrappers
  const actions: Action[] = (manifest.actions ?? []).map((actionDef) => ({
    name: actionDef.name,
    description: actionDef.description,
    similes: actionDef.similes,
    examples: actionDef.examples,

    validate: async (
      _runtime: IAgentRuntime,
      message: Memory,
      state: State | undefined,
    ): Promise<boolean> => {
      const result = exports.validate_action(
        actionDef.name,
        JSON.stringify(message),
        JSON.stringify(state ?? null),
      );
      return typeof result === "boolean" ? result : result !== 0;
    },

    handler: async (
      _runtime: IAgentRuntime,
      message: Memory,
      state: State | undefined,
      options?: HandlerOptions,
      _callback?: HandlerCallback,
    ): Promise<ActionResult> => {
      const resultJson = exports.invoke_action(
        actionDef.name,
        JSON.stringify(message),
        JSON.stringify(state ?? null),
        JSON.stringify(options ?? {}),
      );
      const result: ActionResultPayload = JSON.parse(resultJson);
      return {
        success: result.success,
        text: result.text,
        error: result.error ? new Error(result.error) : undefined,
        data: result.data as Record<string, ProviderValue> | undefined,
        values: result.values as Record<string, ProviderValue> | undefined,
      };
    },
  }));

  // Create provider wrappers
  const providers: Provider[] = (manifest.providers ?? []).map(
    (providerDef) => ({
      name: providerDef.name,
      description: providerDef.description,
      dynamic: providerDef.dynamic,
      position: providerDef.position,
      private: providerDef.private,

      get: async (
        _runtime: IAgentRuntime,
        message: Memory,
        state: State,
      ): Promise<ProviderResult> => {
        const resultJson = exports.get_provider(
          providerDef.name,
          JSON.stringify(message),
          JSON.stringify(state),
        );
        const result: ProviderResultPayload = JSON.parse(resultJson);
        return {
          text: result.text,
          values: result.values as Record<string, ProviderValue> | undefined,
          data: result.data as Record<string, ProviderValue> | undefined,
        };
      },
    }),
  );

  // Create evaluator wrappers
  const evaluators: Evaluator[] = (manifest.evaluators ?? []).map(
    (evalDef) => ({
      name: evalDef.name,
      description: evalDef.description,
      alwaysRun: evalDef.alwaysRun,
      similes: evalDef.similes,
      examples: [] as EvaluationExample[],

      validate: async (
        _runtime: IAgentRuntime,
        message: Memory,
        state: State | undefined,
      ): Promise<boolean> => {
        const result = exports.validate_evaluator(
          evalDef.name,
          JSON.stringify(message),
          JSON.stringify(state ?? null),
        );
        return typeof result === "boolean" ? result : result !== 0;
      },

      handler: async (
        _runtime: IAgentRuntime,
        message: Memory,
        state: State | undefined,
      ): Promise<ActionResult | undefined> => {
        const resultJson = exports.invoke_evaluator(
          evalDef.name,
          JSON.stringify(message),
          JSON.stringify(state ?? null),
        );
        if (!resultJson || resultJson === "null") {
          return undefined;
        }
        const result: ActionResultPayload = JSON.parse(resultJson);
        return {
          success: result.success,
          text: result.text,
          error: result.error ? new Error(result.error) : undefined,
          data: result.data as Record<string, ProviderValue> | undefined,
          values: result.values as Record<string, ProviderValue> | undefined,
        };
      },
    }),
  );

  // Return the plugin
  return {
    name: manifest.name,
    description: manifest.description,
    config: manifest.config ?? {},
    dependencies: manifest.dependencies,
    actions,
    providers,
    evaluators,
    // Routes would need special handling for WASM
    routes: [],
    // Services need special handling
    services: [],

    async init(config: Record<string, string>) {
      exports.init(JSON.stringify(config));
    },
  };
}

/**
 * Preload and validate a WASM plugin without fully loading it
 */
export async function validateWasmPlugin(wasmPath: string): Promise<{
  valid: boolean;
  manifest?: PluginManifest;
  error?: string;
}> {
  try {
    const instance = await loadWasmModule(wasmPath, {});
    const manifestJson = instance.exports.get_manifest();
    const manifest: PluginManifest = JSON.parse(manifestJson);

    // Basic validation
    if (!manifest.name || !manifest.description) {
      return { valid: false, error: "Missing required manifest fields" };
    }

    return { valid: true, manifest };
  } catch (error) {
    return {
      valid: false,
      error: error instanceof Error ? error.message : String(error),
    };
  }
}
